### PR TITLE
Allow sample publication without sending Email

### PR DESCRIPTION
**Pull request by ramonski**
## Description of the issue/feature this PR addresses

This PR allows to publish all the contained samples of publication reports.

## Current behavior before PR

Publication of samples did only happen by sending an email to the contact(s)

## Desired behavior after PR is merged

Publication of samples can be done manually from the Analysis Reports Listing

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: bika/lims/browser/workflow/client.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        transition = transitions.get(status, "prepublish")
+        logger.info("AR Transition: {} -> {}".format(status, transition))
+        try:
+            wf.doActionFor(sample, transition)
+            return True
+        except WorkflowException as e:
+            logger.debug(e)
+            return False
+
+    def add_status_message(self, message, level="info"):
```
Comment (xispa) [inline bika/lims/browser/workflow/client.py]: The super class `RequestContextAware` already provides this function:
https://github.com/senaite/senaite.core/blob/master/bika/lims/browser/workflow/__init__.py#L52-L55

File: bika/lims/browser/workflow/client.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    def __call__(self, action, uids):
+        published = []
+
+        # get the selected ARReport objects
+        reports = map(api.get_object_by_uid, uids)
+        # get all the contained sample UIDs of the generated PDFs
+        sample_uids = map(self.get_sample_uids_in_report, reports)
+        # uniquify the UIDs of the contained samples
+        unique_sample_uids = set(list(
+            itertools.chain.from_iterable(sample_uids)))
```
Comment (xispa) [inline bika/lims/browser/workflow/client.py]: Is there any reason why you use `itertools` here?. I guess you could do `set(sample_uids)`  instead

File: bika/lims/browser/workflow/client.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        transitions = {"verified": "publish",
+                       "published": "republish"}
+        transition = transitions.get(status, "prepublish")
+        logger.info("AR Transition: {} -> {}".format(status, transition))
+        try:
+            wf.doActionFor(sample, transition)
+            return True
+        except WorkflowException as e:
+            logger.debug(e)
+            return False
```
Comment (xispa) [inline bika/lims/browser/workflow/client.py]: Remember we have a `api.get_review_state`. Also, better to use `bika.lims.workflow.doActionFor` instead. This transition will be "promoted" to ancestors and "cascaded" to child, and `doActionFor` from `senaite.core` handles this better. The function will get simpler:
```python

from bika.lims.workflow import wf

def publish_sample(self, sample):
    status = api.get_review_status(sample)
    transitions = {"verified": "publish",  "published": "republish"}
    transition = transitions.get(status, "prepublish")
    succeed, message = wf.doActionFor(sample, transition)
    return succeed
```

File: bika/lims/browser/publish/reports_listing.py [hunk 4]
```diff
@@ -62,10 +62,29 @@ def __init__(self, context, request):
...
+        help_publish_text = _(
+            "Manually publish all contained samples of the selected reports.")
+
+        publish_samples_transition = {
+            "id": "publish_samples",
+            "title": _("Publish"),
+            # see senaite.core.browser.workflow
+            "url": "workflow_action?action=publish_samples",
+            "css_class": "btn-success",
+            "help": help_publish_text,
```
Comment (xispa) [inline bika/lims/browser/publish/reports_listing.py]: Oo, did not know about this "help" attribute!. Awesome!

File: bika/lims/browser/workflow/client.py (hunk 2 above)
Comment (ramonski) [reply bika/lims/browser/workflow/client.py]: The `sample_uids` is actually a list of lists.
For better understanding:
- The users selects 1..n ARReports from the list
- Each report might be a multi-report containing more than one AR/UID
- Of all the selected ARReports, the UIDs need to be extracted -> list of UID lists
- All of these UIDs must be uniquified, looked up and finally transitioned

File: bika/lims/browser/workflow/client.py (hunk 1 above)
Comment (ramonski) [reply bika/lims/browser/workflow/client.py]: well spotted, I'll remove that

File: bika/lims/browser/workflow/client.py (hunk 3 above)
Comment (ramonski) [reply bika/lims/browser/workflow/client.py]: Ok, got it, thanks for the improvement

File: bika/lims/browser/workflow/client.py (hunk 2 above)
Comment (xispa) [reply bika/lims/browser/workflow/client.py]: Aa, I see, thanks for the clarification

**Pull request merged**